### PR TITLE
DNN: add depth2space and space2depth layer for onnx importer

### PR DESCRIPTION
**Pull request by zihaomu**
Hi, I implement the `depth2space` and `space2depth` layers by the `Reshape` and `Permute` Layer, instead of creating a New Layer.

Is it necessary to create a separate layer for `depth2space` and `space2depth`?

[Relate Issue](https://github.com/opencv/opencv/issues/17514)

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 1]
```diff
@@ -2975,6 +2976,85 @@ void ONNXImporter::parseCumSum(LayerParams& layerParams, const opencv_onnx::Node
...
+    // Get blocksize
+    int blocksize = 0;
+    if(layerParams.has("blocksize"))
+        blocksize = layerParams.get<int>("blocksize");
+    CV_CheckGT(blocksize, 0, "");
+
+    // Get mode, only for "DepthToSpace"
+    std::string modeType = "DCR";
+    if(layerParams.has("mode"))
+        modeType = layerParams.get<std::string>("mode");
```
Comment (rogday) [inline modules/dnn/src/onnx/onnx_importer.cpp:2996]:
```cpp
std::string modeType = layerParams.get<std::string>("mode", "DCR");
```

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 2]
```diff
@@ -2975,6 +2976,85 @@ void ONNXImporter::parseCumSum(LayerParams& layerParams, const opencv_onnx::Node
...
+    permuteLp.set("order", DictValue::arrayInt(perm.data(), perm.size()));
+
+    opencv_onnx::NodeProto protoPermute;
+    protoPermute.add_input(reshapeLp.name);
+    protoPermute.add_output(permuteLp.name);
+    addLayer(permuteLp, protoPermute);
+
+    // Step3: Reshape
+    layerParams.type = "Reshape";
+    layerParams.set("dim", DictValue::arrayInt(&shape1[0], shape1.size()));
```
Comment (rogday) [inline modules/dnn/src/onnx/onnx_importer.cpp:3052]: shape1.data()

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 3]
```diff
@@ -2975,6 +2976,85 @@ void ONNXImporter::parseCumSum(LayerParams& layerParams, const opencv_onnx::Node
...
+    opencv_onnx::NodeProto node_proto = node_proto_;
+    const std::string& layer_type = node_proto.op_type();
+    CV_Assert(layer_type == "DepthToSpace" || layer_type == "SpaceToDepth");
+    layerParams.type = layer_type;
+
+    // Get blocksize
+    int blocksize = 0;
+    if(layerParams.has("blocksize"))
+        blocksize = layerParams.get<int>("blocksize");
+    CV_CheckGT(blocksize, 0, "");
```
Comment (rogday) [inline modules/dnn/src/onnx/onnx_importer.cpp:2991]:
```cpp
CV_Assert(layerParams.has("blocksize"));
int blocksize = layerParams.get<int>("blocksize");
```

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 4]
```diff
@@ -2975,6 +2976,85 @@ void ONNXImporter::parseCumSum(LayerParams& layerParams, const opencv_onnx::Node
...
+    MatShape inpShape = outShapes[node_proto.input(0)];
+    CV_Assert(inpShape.size() == 4);
+
+    // Implement DepthToSpace and SpaceToDepth by the Reshape and Permute layer.
+    std::vector<int> shape0, shape1, perm;
+    if(layer_type == "DepthToSpace")
+    {
+        if(modeType == "DCR")
+        {
+            shape0 = {inpShape[0], blocksize, blocksize, inpShape[1]/(blocksize * blocksize), inpShape[2], inpShape[3]};
```
Comment (rogday) [inline modules/dnn/src/onnx/onnx_importer.cpp:3007]: I'm not a fan of inpShape[i], you could name each dimension(N, C, H, W), if you want. But it's fine as is.

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 5]
```diff
@@ -2975,6 +2976,85 @@ void ONNXImporter::parseCumSum(LayerParams& layerParams, const opencv_onnx::Node
...
+    // Get mode, only for "DepthToSpace"
+    std::string modeType = "DCR";
+    if(layerParams.has("mode"))
+        modeType = layerParams.get<std::string>("mode");
+
+    MatShape inpShape = outShapes[node_proto.input(0)];
+    CV_Assert(inpShape.size() == 4);
+
+    // Implement DepthToSpace and SpaceToDepth by the Reshape and Permute layer.
+    std::vector<int> shape0, shape1, perm;
```
Comment (rogday) [inline modules/dnn/src/onnx/onnx_importer.cpp:3002]:
```cpp
std::array<int, 6> shape0, perm;
std::array<int, 4> shape1;
```

**Review by rogday — CHANGES_REQUESTED** [review]
Thank you for your contribution! Nope, it isn't necessary.

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 5 above)
Comment (zihaomu) [reply modules/dnn/src/onnx/onnx_importer.cpp]: Thanks for the code review. All comments mentioned issues have been fixed.

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 6]
```diff
@@ -2975,6 +2976,84 @@ void ONNXImporter::parseCumSum(LayerParams& layerParams, const opencv_onnx::Node
...
+    // Get blocksize
+    CV_Assert(layerParams.has("blocksize"));
+    int blocksize = layerParams.get<int>("blocksize");
+
+    // Get mode, only for "DepthToSpace"
+    std::string modeType = layerParams.get<std::string>("mode", "DCR");
+
+    MatShape inpShape = outShapes[node_proto.input(0)];
+    int N = inpShape[0], C = inpShape[1], H = inpShape[2], W = inpShape[3];
+    CV_Assert(inpShape.size() == 4);
```
Comment (rogday) [inline modules/dnn/src/onnx/onnx_importer.cpp:2996]: We should swap the order of these lines, otherwise it might be UB.

Comment (zihaomu) [reply modules/dnn/src/onnx/onnx_importer.cpp:2996]: It's My fault. Thanks, fixed.

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 7]
```diff
@@ -2975,6 +2976,84 @@ void ONNXImporter::parseCumSum(LayerParams& layerParams, const opencv_onnx::Node
...
+
+    // Implement DepthToSpace and SpaceToDepth by the Reshape and Permute layer.
+    std::array<int, 6> shape0, perm;
+    std::array<int, 4> shape1;
+
+    if(layer_type == "DepthToSpace")
+    {
+        if(modeType == "DCR")
+        {
+            shape0 = {N, blocksize, blocksize, C/(blocksize * blocksize), H, W};
```
Comment (alalek) [inline modules/dnn/src/onnx/onnx_importer.cpp:3006]: `blockSize` is assumed to be non-zero. Need to validate that (validate inputs).

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 8]
```diff
@@ -2975,6 +2976,84 @@ void ONNXImporter::parseCumSum(LayerParams& layerParams, const opencv_onnx::Node
...
+        }
+        else if (modeType == "CRD")
+        {
+            shape0 = {N, C/(blocksize * blocksize), blocksize, blocksize, H, W};
+            perm = {0, 1, 4, 2, 5, 3};
+            shape1 = {N, C/(blocksize * blocksize), H * blocksize, W * blocksize};
+        }
+        else
+            CV_Error(Error::StsNotImplemented, "The mode of"+ modeType + " in DepthToSpace Layer is not supported");
+    }else // SpaceToDepth
```
Comment (alalek) [inline modules/dnn/src/onnx/onnx_importer.cpp:3018]: coding style - `else` from new line

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 9]
```diff
@@ -2975,6 +2976,84 @@ void ONNXImporter::parseCumSum(LayerParams& layerParams, const opencv_onnx::Node
...
+    int N = inpShape[0], C = inpShape[1], H = inpShape[2], W = inpShape[3];
+    CV_Assert(inpShape.size() == 4);
+
+    // Implement DepthToSpace and SpaceToDepth by the Reshape and Permute layer.
+    std::array<int, 6> shape0, perm;
+    std::array<int, 4> shape1;
+
+    if(layer_type == "DepthToSpace")
+    {
+        if(modeType == "DCR")
```
Comment (alalek) [inline modules/dnn/src/onnx/onnx_importer.cpp:3004]: coding style - keep space after `if`

**Review by rogday — APPROVED** [review]
LGTM! :+1:

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 10]
```diff
@@ -2975,6 +2976,84 @@ void ONNXImporter::parseCumSum(LayerParams& layerParams, const opencv_onnx::Node
     addLayer(layerParams, node_proto);
 }
 
+void ONNXImporter::parseDepthToSpace(LayerParams& layerParams, const opencv_onnx::NodeProto& node_proto_)
+{
+    // We parse "DepthToSpace" and "SpaceToDepth" in this function.
+    opencv_onnx::NodeProto node_proto = node_proto_;
+    const std::string& layer_type = node_proto.op_type();
+    CV_Assert(layer_type == "DepthToSpace" || layer_type == "SpaceToDepth");
+    layerParams.type = layer_type;
```
Comment (alalek) [inline modules/dnn/src/onnx/onnx_importer.cpp:2985]: > layerParams.type = layer_type;

Is it used anywhere?
Below its value is rewritten.

Comment (zihaomu) [reply modules/dnn/src/onnx/onnx_importer.cpp]: Thanks for code reviewing. All issues have been fixed.

**Comment by asmorkalov** [comment]
@zihaomu Friendly reminder. Also please rebase the branch on top of upstream 4.x to enable new CI checks.

**Comment by zihaomu** [comment]
> @zihaomu Friendly reminder. Also please rebase the branch on top of upstream 4.x to enable new CI checks.

Thanks for reminding. I will rebase it now.

**Review by rogday — APPROVED** [review]
:+1:

**Pull request merged**